### PR TITLE
cpu: support for detecting nx-gzip coprocessor feature

### PR DESCRIPTION
**Pull request by Chandan-Abhyankar**
Signed-off-by: Chandan Abhyankar <Chandan.Abhyankar@ibm.com>
Adding nx-gzip label support

### CODE REVIEW

**Comment by prb112** [comment]
Hi @adrianchiris @Ethyling 
I hope all is well. 
Are there any steps we can take to address your comments/concerns? 
Many thanks
Paul

**Comment by Ethyling** [comment]
Hey @Chandan-Abhyankar  @prb112. Is this feature available for every archs or should we move this to an arch dependent file with a stub file for archs that don't support it? Something similar to [rdt_amd64.go](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/source/cpu/rdt_amd64.go) and [rdt_stub.go](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/source/cpu/rdt_stub.go). I don't have a lot of knowledge about CPU attributes. Otherwise the code looks good to me, but I don't have the hardware to test it

**Comment by prb112** [comment]
Hey @Ethyling it is architecture specific. We could move it to a separate file.  Do you want us to take that step? Many thanks, Paul

**Comment by Ethyling** [comment]
@prb112 Yes, I think we should move this to a separate file so we don't include it for archs that don't have this feature. Thank you

**Comment by prb112** [comment]
> @prb112 Yes, I think we should move this to a separate file so we don't include it for archs that don't have this feature. Thank you

We'll take care of it. Thanks @Ethyling

**Comment by Chandan-Abhyankar** [comment]
Hi @Ethyling As changes are architecture specific [Power], I have moved code to cpuid_ppc64le.go.
Kindly have a re-look and let me know if you have any comments. 
Thank you

**Comment by Ethyling** [comment]
Hey @Chandan-Abhyankar, I don't this this feature is related to `cpuid` so I thinkg we should move it to its own file (something like `nx_ppc64le.go`). We should also provide a stub file (i.e. `nx_stub.go`) for the other archs

**Comment by prb112** [comment]
Hi @Ethyling 

[nx-gzip](https://files.openpower.foundation/s/ZmtZyCGiJ2oJHim) is an on-chip accelerator to facilitate compression/decompression on PowerPC.  [PR from centos](https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/merge_requests/737) elaborates in more detail. 

Do you still want us to split this out? 

Thank you for the clarification, 

Paul

**Comment by Ethyling** [comment]
Hi @prb112, yes I still think we should split it in its own file. Let's ask inputs from @marquiz to confirm this

**Comment by prb112** [comment]
> Hi @prb112, yes I still think we should split it in its own file. Let's ask inputs from @marquiz to confirm this

Thanks, I'll wait on the confirmation, take care, Paul

**Comment by yselkowitz** [comment]
CPUID should be just that, which this isn't.  It should be split out and handled separately (but still under `source/cpu`), like `pstate` or `security`.

Also, be sure to use `hostpath.SysfsDir.Path` as elsewhere in NFD.

**Comment by prb112** [comment]
Thank you for the feedback @yselkowitz @marquiz we've taken the comments into account and are working on a revision.

**Comment by Chandan-Abhyankar** [comment]
@yselkowitz @Ethyling As suggested, I have added new file nx_ppc64le.go inside source/cpu to handle NX-GZIP support for Power. Kindly have a re-look and let us know incase of any comments.

**Review by marquiz — COMMENTED** [review]
Thanks @Chandan-Abhyankar. I started to think could we come up with some common "title" for these kind of accelearator features. Similar to what we did for security in #833. I.e. instead of `nx.nx_gzip` have like `coprocessor.nx_gzip` or `accelerator.nx_gzip` or `crypto.nx_gzip` or smth. That would reduce clutter when/if adding other similar features in the future. WDYT?

In any case we need to also update the documentation accordingly, explaining the new feature (`docs/usage/features.md` and `docs/usage/customization-guide.md`).

**Comment by prb112** [comment]
> `coprocessor`

Hey @marquiz 

Thank you for the comments.

This does make a bit of sense. NX-GZIP is technically a coprocessor. 

Netting the comments to actions: 
1. Change nx_stub.go nx_ppc64le.go to coprocessor_stub.go coprocess_ppc64le.go 
2. Change to a generic coprocessor
```s.features.Attributes[NXFeature] = nfdv1alpha1.NewAttributeFeatures(discoverNxGzip())```
to 
```s.features.Attributes[NXFeature] = nfdv1alpha1.NewAttributeFeatures(discoverCoprocessor())```
3. Update the doc in `docs/usage/features.md` and `docs/usage/customization-guide.md`

Thanks for taking the time to confirm

Paul

**Comment by Chandan-Abhyankar** [comment]
@marquiz @prb112 As per your suggestions, I have incorporated review comments regarding coprocessor feature(nx-gzip). Kindly have a look and let us know in-case of any comments

**Comment by prb112** [comment]
Hi @marquiz I hope all is well. 

Are the changes @Chandan-Abhyankar made as expected? 

We look forward to the review. 

Best wishes

Paul

File: docs/usage/features.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ option of nfd-worker.
 | ----------------------- | ------------ | -----------
 | **`cpu-cpuid.<cpuid-flag>`**      | true   | CPU capability is supported. **NOTE:** the capability might be supported but not enabled.
 | **`cpu-hardware_multithreading`** | true   | Hardware multithreading, such as Intel HTT, enabled (number of logical CPUs is greater than physical CPUs)
+| **`cpu-nx_gzip`**                 | true   | Nest Accelerator for GZIP is supported(Power).
```
Comment (marquiz) [inline docs/usage/features.md:51]: I think we could use the same `coprocessor` "sub-cpu-feature-prefix" as for the raw feature.

```suggestion
| **`cpu-coprocessor.nx_gzip`**                 | true   | Nest Accelerator for GZIP is supported(Power).
```

File: source/cpu/coprocessor_ppc64le.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+//go:build ppc64le
+// +build ppc64le
+
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (marquiz) [inline source/cpu/coprocessor_ppc64le.go:5]: nit
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: source/cpu/coprocessor_stub.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+//go:build !ppc64le
+// +build !ppc64le
+
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (marquiz) [inline source/cpu/coprocessor_stub.go:5]: nit
```suggestion
Copyright 2023 The Kubernetes Authors.
```

**Review by marquiz — COMMENTED** [review]
Thanks for the update @Chandan-Abhyankar 👍

> Are the changes @Chandan-Abhyankar made as expected?

Woops, I seem to have dropped the ball on this one 🙄

I had just a few small comments but otherwise looks good.

**Review by marquiz — COMMENTED** [review]
Thanks @Chandan-Abhyankar for the update, looks good to me now. One more ask, could you squash the changes into one commit?

**Comment by Chandan-Abhyankar** [comment]
@marquiz - Thanks for review and comments. I have made all the changes. Do you want me to discard all changes and include them back in 1 commit?

**Comment by marquiz** [comment]
> Thanks for review and comments. I have made all the changes. Do you want me to discard all changes and include them back in 1 commit?

Yeah. Or well, not discard the changes but "discard" the extra commits with e.g. `git rebase -i origin/master`, squash all extra commits into the first one with `fixup` and then force push.

Also, I think you could improve the commit message a bit (see [K8s commit message guidelines](https://www.kubernetes.dev/docs/guide/pull-requests/#commit-message-guidelines) for inspiration). Change the commit msg subject to e.g.`"Add support for detecting nx-gzip coprocessor feature"` or `"cpu: support for detecting nx-gzip coprocessor feature"`. Then in the commit message body you could shortly spell out some details, e.g. that this is a accelerator on ibm-z systems or smth.

**Comment by Chandan-Abhyankar** [comment]
@marquiz , Thanks a lot for your pointers, help and support. I have now integrated all changes in one commit. Kindly have a re-look.

**Comment by marquiz** [comment]
> Thanks a lot for your pointers, help and support. I have now integrated all changes in one commit. Kindly have a re-look.

Thanks @Chandan-Abhyankar. Still one nit about the commit message body 🙈 Use full sentences, starting with a capital letter and end with a period, e.g. `Nest accelerator gzip support for IBM Power systems.`

You could also remove the merge commit by rebasing on top of latest master.

**Comment by Chandan-Abhyankar** [comment]
@marquiz , Kindly review changes. I was facing issues with rebasing on top of latest master, hence followed my usual steps to check-in code. 
For features.md and customization-guide.md (inside docs/usage folder), I took latest code from master branch and appended my changes. Hope all is good now. Thanks

**Comment by marquiz** [comment]
Thanks @Chandan-Abhyankar for the update. The commit looks a bit strange as it touches unrelated files, in docs for example. Could you do `git fetch origin && git rebase origin/master` and then re-push to base this on the latest master? Other than that, looks good now

**Review by marquiz — APPROVED** [review]
Thanks @Chandan-Abhyankar, looks good to go to me 👍

**Comment by marquiz** [comment]
/retitle cpu: support for detecting nx-gzip coprocessor feature

**Comment by Chandan-Abhyankar** [comment]
@marquiz , Thanks a lot for your help and support.

**Comment by prb112** [comment]
Thank you all - we appreciate all your review and effort in supporting this feature.

**Pull request merged**